src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandler.scala

```diff
@@ -0,0 +1,446 @@
+  }
+
+  private def handleBlockHeaders(syncState: SyncState, peer: Peer, headers: Seq[BlockHeader]): (FastSyncCommand, SyncState) = {
+    processHeaders(syncState, peer, headers) match {
+      case ParentDifficultyNotFound(header, finalState) =>
+        log.debug("Parent difficulty not found for block {}, not processing rest of headers", header.number)
+        (ContinueSyncing(None), finalState)
+      case HeadersProcessingFinished(finalState) =>
+        (ContinueSyncing(None), finalState)
+      case ImportedTargetBlock(finalState)  =>
+        (InitTargetBlockUpdate(ImportedLastBlock), finalState)
+      case ValidationFailed(header, peerToBlackList, finalState) =>
+        (handleBlockValidationError(finalState, header, peerToBlackList), finalState)
+    }
+  }
+
+  @tailrec
+  private def processHeaders(syncState: SyncState, peer: Peer, headers: Seq[BlockHeader]): HeaderProcessingResult = {
+    if (headers.nonEmpty) {
+      val header = headers.head
+      processHeader(syncState, header, peer) match {
+        case Left(result)        => result
+        case Right(headerAndDif) =>
+          val newState = updateSyncState(headerAndDif._1, headerAndDif._2)
+          if (header.number == syncState.safeDownloadTarget){
+            ImportedTargetBlock(newState)
+          } else {
+            processHeaders(newState ,peer, headers.tail)
+          }
+      }
+    } else
+      HeadersProcessingFinished(syncState)
+  }
+
+  private def processHeader(syncState: SyncState, header: BlockHeader, peer: Peer) : Either[HeaderProcessingResult , (ValidationSucces, BigInt)] = for {
+    validatedHeader  <- validateHeader(syncState, header, peer)
+    parentDifficulty <- blockchain.getTotalDifficultyByHash(header.parentHash).toRight(ParentDifficultyNotFound(header, validatedHeader.newState))
+  } yield (validatedHeader, parentDifficulty)
+
+  private def validateHeader(syncState: SyncState, header: BlockHeader, peer: Peer): Either[ValidationFailed, ValidationSucces] = {
+    import syncConfig.{fastSyncBlockValidationK => K, fastSyncBlockValidationX => X, fastSyncBlockValidationN => N}
+    val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
+
+    if (shouldValidate) {
+      validators.blockHeaderValidator.validate(header, blockchain.getBlockHeaderByHash) match {
+        case Right(_) =>
+          Right(ValidationSucces(header, syncState.updateNextBlockToValidate(header, K, X, randomFun)))
+
+        case Left(error) =>
+          log.debug(s"Block header validation failed during fast sync at block ${header.number}: $error")
+          Left(ValidationFailed(header, peer, syncState.updateDiscardedBlocks(header, N)))
+      }
+    } else {
+      Right(ValidationSucces(header, syncState))
+    }
+  }
+
+  private def updateSyncState(validatedHeader: ValidationSucces, parentTd: BigInt): SyncState = {
+    blockchain.save(validatedHeader.header)
```

Comment:
maybe consider extracting `validatedHeader.header` and `validatedHeader.header.hash` because of many calls also you're only using `validatedHeader.header` so maybe passing `ValidationSucces` object is to much

Reply:
i can extract `validatedHeader.header` to some local variable `header`, but i like passing `ValidationSuccess` as type immediately tell that header was already successfully validated.